### DISCUSSION TITLE
How do I get all available namespaces?

### DISCUSSION THREAD

**Discussion by madelaney** [kube-rs/kube#1344] [Q&A] [answered]
I was looking over the pod example:
```rust
  let pods: Api<Pod> = Api::default_namespaced(client);
  for p in pods.list(&ListParams::default()).await? {
    info!("found pod {}", p.name_any());
  }
```

and I was curious if there was a way to get all of the namespaces in a similar fashion?

**Comment by clux** [comment, marked as answer]
make an `Api` on [`Namespace`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.Namespace.html)
```rust
let ns: Api<Namepace> = Api::all(client)
ns.list(...).await
```

**Reply by madelaney** [reply]
Oh that's awesome. Thank you for the example, and the quick reply.

**Reply by LennonChin** [reply]
but how do i get multiple specified namespaces?

**Reply by clux** [reply]
you can loop over the namespaces you want and create a `Api::namespaced(ns_i, ...)` inside the loop

...or, if you are just listing/watching from multiple given namespaces you can technically do a single `Api::all` and use do a `api.list` with a specific field selector (on `ListParams`) with "metadata.namespace=ns1,metadata.namespace=ns2,...". (that doesn't work for `.get` though - then you need an `Api::namespaced`)

EDIT: you can only list out the complement turns out. see [optimization guide for selectors](https://kube.rs/controllers/optimization/#reducing-number-of-watched-objects)

**Reply by LennonChin** [reply]
Thanks for your help! Will this method of `Api::all` combined with`ListParams` put pressure on APIServer? Will it only read the namespace data specified in`ListParams` from APIServer, or read all namespace data and filter it on the client side?

**Reply by clux** [reply]
field selectors or label selectors will do filter-equivalents server-side, so should be better when possible.

the `Api` construct (and associated params) are meant to model what the apiserver supports so ~WYSIWYG. (conversely; `jsonpath` [must be done client-side](https://github.com/kube-rs/kube/blob/c7054b5c86d09947cc4dce210cd24d8487b16f2e/examples/dynamic_jsonpath.rs#L18-L39) - you cannot specify jsonpath queries on `ListParams` because it's not part of the api)

**Reply by LennonChin** [reply]
Thank you for your answer, it's great!

**Comment by LennonChin** [comment]
> you can loop over the namespaces you want and create a `Api::namespaced(ns_i, ...)` inside the loop
> 
> ...or, if you are just listing/watching from multiple given namespaces you can technically do a single `Api::all` and use do a `api.list` with a specific field selector (on `ListParams`) with "metadata.namespace=ns1,metadata.namespace=ns2,...". (that doesn't work for `.get` though - then you need an `Api::namespaced`)

Unfortunately, after testing, `list` and `watch` do not support field selectors with multiple namespaces, and client-go does not support it either.

**Reply by SOF3** [reply]
Also note that (iirc) `Api::all` with field selectors requires a ClusterRole giving you access to list, while field selectors only require list access to a particular namespace, so you probably wanted to just do multiple `Api::namespaced()` anyway

**Reply by clux** [reply]
> list and watch do not support field selectors with multiple namespaces

bah, i didn't know it didn't support listing the namespaces, because i use it for listing the complement :(

..you have to either exclude ala `kubectl get svc -A --field-selector "metadata.namespace!=default,metadata.namespace!=kube-system"` or build labels instead

have updated the [optimization guide selector section](https://kube.rs/controllers/optimization/#reducing-number-of-watched-objects) to reflect this :(
